gamemodes/terrortown/gamemode/shared/sh_sprint.lua

```diff
@@ -44,31 +44,54 @@ if SERVER then
 		ply.oldSprintProgress = ply.sprintProgress
 		ply.sprintMultiplier = bool and (1 + maxSprintMul:GetFloat()) or nil
 		ply.isSprinting = bool
-		ply.sprintTS = CurTime()
 	end)
 else
-	local function PlayerSprint(bool)
+	local lastPress = 0
+
+	local function PlayerSprint(trySprinting, bind)
 		local client = LocalPlayer()
 
-		if bool and not GetGlobalBool("ttt2_sprint_enabled", true) or not bool and not client.isSprinting then return end
+		if trySprinting and not GetGlobalBool("ttt2_sprint_enabled", true) then return end
+		if not trySprinting and not client.isSprinting or trySprinting and client.isSprinting then return end
+		if client.isSprinting and (client.useSprintBind and not bind or not client.useSprintBind and bind) then return end
 
 		client.oldSprintProgress = client.sprintProgress
-		client.sprintMultiplier = bool and (1 + GetGlobalFloat("ttt2_sprint_max", 0)) or nil
-		client.isSprinting = bool
-		client.sprintTS = CurTime()
+		client.sprintMultiplier = trySprinting and (1 + GetGlobalFloat("ttt2_sprint_max", 0)) or nil
+		client.isSprinting = trySprinting
+		client.useSprintBind = bind
 
 		net.Start("TTT2SprintToggle")
-		net.WriteBool(bool)
+		net.WriteBool(trySprinting)
 		net.SendToServer()
 	end
 
+	hook.Add("KeyPress", "TTT2DoublePressSprint", function(ply, key)
+		if not IsFirstTimePredicted() then return end
+		if key == IN_FORWARD and not ply.isSprinting then
+			local time = CurTime()
```

Comment:
I don't gave a deep look, I think CurTime is correct because this hook should get triggered as soon as I press a key (not any frame / tick). But just to think about, it's correct not using FrameTime, right?

Reply:
Yes, the hook is triggered on key down events. I use CurTime to check if the button was pressed twice in the last 0.4 seconds. In this case CurTime makes most sense, since the button presses could be multiple frames apart and still count as a double tap :)